kythe/go/serving/graph/columnar/columnar_encoding.go

```diff
@@ -0,0 +1,177 @@
+/*
+ * Copyright 2018 The Kythe Authors. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this src except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// Package columnar implements the columnar table format for a Kythe xrefs service.
+package columnar
+
+import (
+	"fmt"
+
+	"kythe.io/kythe/go/util/keys"
+
+	"github.com/golang/protobuf/proto"
+
+	gspb "kythe.io/kythe/proto/graph_serving_go_proto"
+	scpb "kythe.io/kythe/proto/schema_go_proto"
+	spb "kythe.io/kythe/proto/storage_go_proto"
+)
+
+var (
+	// EdgesKeyPrefix is the common key prefix for all Kythe columnar Edges
+	// key-value entries.
+	EdgesKeyPrefix, _ = keys.Append(nil, "eg")
+)
+
+func init() {
+	// Restrict the capacity of the key prefix to ensure appending to it creates a new array.
+	EdgesKeyPrefix = EdgesKeyPrefix[:len(EdgesKeyPrefix):len(EdgesKeyPrefix)]
+}
+
+// Columnar edges group numbers.
+// See: kythe/proto/graph_serving.proto
+const (
+	columnarEdgesIndexGroup  = -1 // no group number
+	columnarEdgesEdgeGroup   = 10
+	columnarEdgesTargetGroup = 20
+)
+
+// KV is a single columnar key-value entry.
+type KV struct{ Key, Value []byte }
+
+// EncodeEdgesEntry encodes a columnar Edges entry.
+func EncodeEdgesEntry(keyPrefix []byte, eg *gspb.Edges) (*KV, error) {
+	switch e := eg.Entry.(type) {
+	case *gspb.Edges_Index_:
+		return encodeEdgesIndex(keyPrefix, eg.Source, e.Index)
+	case *gspb.Edges_Edge_:
+		return encodeEdge(keyPrefix, eg.Source, e.Edge)
+	case *gspb.Edges_Target_:
+		return encodeEdgesTarget(keyPrefix, eg.Source, e.Target)
+	default:
+		return nil, fmt.Errorf("unknown Edges entry: %T", e)
+	}
+}
+
+func encodeEdgesIndex(prefix []byte, src *spb.VName, idx *gspb.Edges_Index) (*KV, error) {
+	key, err := keys.Append(prefix, src)
+	if err != nil {
+		return nil, err
+	}
+	val, err := proto.Marshal(idx)
+	if err != nil {
+		return nil, err
+	}
+	return &KV{key, val}, nil
+}
+
+func encodeEdge(prefix []byte, src *spb.VName, e *gspb.Edges_Edge) (*KV, error) {
+	key, err := keys.Append(prefix, src, columnarEdgesEdgeGroup,
+		e.GetGenericKind(), int32(e.GetKytheKind()), e.Ordinal, e.Reverse, e.Target)
```

Comment:
What is the type of e.GetKytheKind() that we have to cast it to int32? If it is int64 why don't we just store the int64?

Reply:
It's an enum: https://github.com/google/kythe/blob/fe5b532f75f08e1847483b566b6302da1a7a4ab0/kythe/proto/schema_go_proto/schema.pb.go#L262.  It must be cast for the `keys` package .